packages/expo/src/executors/download/download.impl.ts

```diff
@@ -0,0 +1,136 @@
+  options: ExpoEasDownloadOptions,
+  context: ExecutorContext
+): AsyncGenerator<ReactNativeDownloadOutput> {
+  const projectRoot = context.workspace.projects[context.projectName].root;
+  ensureNodeModulesSymlink(context.root, projectRoot);
+
+  const build = getBuild(context.root, projectRoot, options);
+  const buildUrl = build?.artifacts?.buildUrl;
+  if (!buildUrl) {
+    throw new Error(`No build URL found.`);
+  }
+
+  if (!existsSync(options.output)) {
+    mkdirSync(options.output, { recursive: true });
+  }
+
+  const downloadFileName = buildUrl.split('/').pop();
+  const downloadFilePath = join(options.output, downloadFileName);
+  await downloadBuild(buildUrl, downloadFilePath);
+
+  const appExtension = getAppExtension(build.platform, downloadFileName);
+  const appName = `${names(build.project?.name).className}${appExtension}`;
+  const outputFilePath = join(options.output, appName);
+
+  if (downloadFileName.endsWith('.tar.gz')) {
+    await unzipBuild(downloadFilePath, options.output);
+  } else {
+    await copyBuildFile(downloadFilePath, outputFilePath);
+  }
+
+  logger.info(`Succesfully download the build to ${outputFilePath}`);
+
+  yield { success: true };
+}
+
+async function downloadBuild(buildUrl: string, output: string) {
+  const streamPipeline = promisify(pipeline);
+
+  const response = await fetch(buildUrl);
+
+  if (!response.ok)
+    throw new Error(
+      `Unable to download the build ${buildUrl}. Error: ${response.statusText}`
+    );
+
+  await streamPipeline(response.body, createWriteStream(output));
+}
+
+export function getBuild(
+  workspaceRoot: string,
+  projectRoot: string,
+  options: ExpoEasDownloadOptions
+) {
+  const buildList = runCliBuildList(workspaceRoot, projectRoot, {
+    ...options,
+    json: true,
+    status: 'finished',
+    limit: 1,
+  });
+  const builds = JSON.parse(buildList);
+  if (!builds.length) {
+    throw new Error(
+      `No EAS build found. Please check expo.dev to make sure your build is finished.`
+    );
+  }
+  logger.info(`${chalk.bold.cyan('info')} Found build: ${buildList}`);
+
+  return builds[0];
+}
+
+export function getAppExtension(
+  platform: string,
+  downloadFileName: string
+): string {
+  platform = platform.toLowerCase();
+  if (platform === 'ios') {
+    return '.app';
+  }
+  if (downloadFileName.includes('.')) {
+    return downloadFileName.split('.').pop();
+  }
+}
+
+export async function unzipBuild(
+  downloadFilePath: string,
+  outputDirectoryPath: string
+) {
+  const streamPipeline = promisify(pipeline);
+  const unzip = createUnzip();
+  const extract = tar.extract(outputDirectoryPath);
+  await streamPipeline(createReadStream(downloadFilePath), unzip, extract);
+}
+
+export function copyBuildFile(
+  downloadFilePath: string,
+  outputFilePath: string
+) {
+  return new Promise<void>((resolve, reject) => {
+    copyFile(downloadFilePath, outputFilePath, (error) => {
+      if (error) {
+        reject(error);
+      }
```

Comment:
Probably should wrap resolve in else block. Technically this is fine, but the error shouldn't result in calling resolve after reject.